db/forward_iterator.cc

```diff
@@ -211,7 +211,8 @@ class ForwardLevelIterator : public InternalIterator {
   Status status_;
   InternalIterator* file_iter_;
   PinnedIteratorsManager* pinned_iters_mgr_;
-  const SliceTransform* prefix_extractor_;
+  // Kept alive by ForwardIterator::sv_->mutable_cf_options
+  const std::shared_ptr<const SliceTransform>& prefix_extractor_;
```

Comment:
AFAIK, SuperVersion::mutable_cf_options, Version::mutable_cf_options_ and Compaction::mutable_cf_options_ are actually immutable. For this change, they are generally referenced in a context where something is holding on the that owning entity. For example, if an iterator doesn't hold on to a SuperVersion, it can't guarantee other data stays around.

The UAF case I ran into above as an iterator refresh. I should probably look for more similar to that.

Copying shared_ptr in too many places would likely negate most of the performance gains.

Reply:
> The UAF case I ran into above is an iterator refresh. I should probably look for more similar to that.

I inspected ArenaWrappedDBIter::Refresh and it looks safe. It re-loads everything including prefix_extractor after changing SuperVersion.